File: android-app/app/src/main/java/com/example/missedcallsms/MainActivity.java
package com.example.missedcallsms;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.PowerManager;
import android.provider.Settings;
import android.telephony.SubscriptionInfo;
import android.telephony.SubscriptionManager;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final int REQ_PERMISSIONS = 1;

    private Prefs prefs;
    private Switch enableSwitch;
    private Spinner simSpinner;
    private EditText messageEdit;
    private TextView statusText;
    private TextView batteryWarning;
    private Button batteryBtn;

    private final List<Integer> subscriptionIds = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        prefs = new Prefs(this);

        enableSwitch = findViewById(R.id.switch_enable);
        simSpinner = findViewById(R.id.spinner_sim);
        messageEdit = findViewById(R.id.edit_message);
        statusText = findViewById(R.id.text_status);
        batteryWarning = findViewById(R.id.text_battery_warning);
        batteryBtn = findViewById(R.id.btn_battery);
        Button grantBtn = findViewById(R.id.btn_grant);
        Button saveBtn = findViewById(R.id.btn_save);

        enableSwitch.setChecked(prefs.isEnabled());
        messageEdit.setText(prefs.getMessage());

        grantBtn.setOnClickListener(v -> requestPermissions());
        saveBtn.setOnClickListener(v -> saveSettings());
        batteryBtn.setOnClickListener(v -> requestBatteryOptimizationExemption());

        updateStatusText();
        updateBatteryWarning();
        populateSimSpinner();
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateStatusText();
        updateBatteryWarning();
    }

    private void updateBatteryWarning() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
            if (!pm.isIgnoringBatteryOptimizations(getPackageName())) {
                batteryWarning.setText(
                    "Battery optimization is ON — Android may block calls from being detected. Tap below to fix.");
                batteryWarning.setVisibility(View.VISIBLE);
                batteryBtn.setVisibility(View.VISIBLE);
            } else {
                batteryWarning.setVisibility(View.GONE);
                batteryBtn.setVisibility(View.GONE);
            }
        }
    }

    private void requestBatteryOptimizationExemption() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Intent intent = new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
            intent.setData(Uri.parse("package:" + getPackageName()));
            startActivity(intent);
        }
    }

    private void populateSimSpinner() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE)
                != PackageManager.PERMISSION_GRANTED) {
            simSpinner.setEnabled(false);
            return;
        }

        SubscriptionManager sm = getSystemService(SubscriptionManager.class);
        List<SubscriptionInfo> sims = null;
        try {
            sims = sm.getActiveSubscriptionInfoList();
        } catch (Exception ignored) {}

        subscriptionIds.clear();
        List<String> labels = new ArrayList<>();

        if (sims != null && !sims.isEmpty()) {
            for (SubscriptionInfo info : sims) {
                CharSequence name = info.getDisplayName();
                boolean isEsim = Build.VERSION.SDK_INT >= 28 && info.isEmbedded();
                int slot = info.getSimSlotIndex();

                String label;
                if (name != null && name.length() > 0) {
                    label = name.toString();
                } else if (isEsim) {
                    label = "eSIM";
                } else {
                    label = "SIM " + (slot + 1);
                }
                if (isEsim) label = "[eSIM] " + label;

                String number = info.getNumber();
                if (number != null && !number.isEmpty()) label += " (" + number + ")";
                labels.add(label);
                subscriptionIds.add(info.getSubscriptionId());
            }
        } else {
            int defaultSubId = SubscriptionManager.getDefaultSmsSubscriptionId();
            labels.add("Default SIM");
            subscriptionIds.add(defaultSubId != SubscriptionManager.INVALID_SUBSCRIPTION_ID
                ? defaultSubId : -1);
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this,
            android.R.layout.simple_spinner_item, labels);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        simSpinner.setAdapter(adapter);
        simSpinner.setEnabled(true);

        int savedSubId = prefs.getSubscriptionId();
        int savedIndex = subscriptionIds.indexOf(savedSubId);
        if (savedIndex >= 0) simSpinner.setSelection(savedIndex);

        simSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override public void onItemSelected(AdapterView<?> p, View v, int pos, long id) {}
            @Override public void onNothingSelected(AdapterView<?> p) {}
        });
    }

    private void saveSettings() {
        prefs.setEnabled(enableSwitch.isChecked());
        prefs.setMessage(messageEdit.getText().toString().trim());
        int selectedPos = simSpinner.getSelectedItemPosition();
        if (selectedPos >= 0 && selectedPos < subscriptionIds.size()) {
            prefs.setSubscriptionId(subscriptionIds.get(selectedPos));
        }
        Toast.makeText(this, "Settings saved.", Toast.LENGTH_SHORT).show();
        updateStatusText();
    }

    private void requestPermissions() {
        List<String> missing = new ArrayList<>();
        String[] base = {
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.READ_CALL_LOG,
            Manifest.permission.SEND_SMS,
            Manifest.permission.READ_PHONE_NUMBERS
        };
        for (String p : base) {
            if (ContextCompat.checkSelfPermission(this, p) != PackageManager.PERMISSION_GRANTED)
                missing.add(p);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS)
                    != PackageManager.PERMISSION_GRANTED)
                missing.add(Manifest.permission.POST_NOTIFICATIONS);
        }
        if (missing.isEmpty()) {
            Toast.makeText(this, "All permissions already granted.", Toast.LENGTH_SHORT).show();
            populateSimSpinner();
        } else {
            ActivityCompat.requestPermissions(this, missing.toArray(new String[0]), REQ_PERMISSIONS);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQ_PERMISSIONS) {
            boolean allGranted = true;
            for (int r : grantResults) {
                if (r != PackageManager.PERMISSION_GRANTED) { allGranted = false; break; }
            }
            if (allGranted) {
                Toast.makeText(this, "All permissions granted!", Toast.LENGTH_SHORT).show();
                populateSimSpinner();
            } else {
                Toast.makeText(this,
                    "Some permissions denied. App may not work correctly.", Toast.LENGTH_LONG).show();
            }
            updateStatusText();
        }
    }

    private void updateStatusText() {
        boolean hasSms = ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
            == PackageManager.PERMISSION_GRANTED;
        boolean hasPhone = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE)
            == PackageManager.PERMISSION_GRANTED;
        boolean hasCallLog = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CALL_LOG)
            == PackageManager.PERMISSION_GRANTED;

        if (!hasPhone || !hasSms || !hasCallLog) {
            statusText.setText("Permissions needed \u2014 tap Grant Permissions");
        } else if (!prefs.isEnabled()) {
            statusText.setText("Auto-reply is OFF. Toggle the switch and save.");
        } else {
            statusText.setText("Active \u2014 will auto-reply to missed calls.");
        }
    }
}

File: MissedCallReply/app/src/main/java/com/personalapp/missedcallreply/MainActivity.kt
package com.personalapp.missedcallreply

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.telephony.SubscriptionInfo
import android.telephony.SubscriptionManager
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.ContextCompat.startForegroundService
import com.personalapp.missedcallreply.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var simList: List<SubscriptionInfo> = emptyList()

    private val permissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { results ->
        if (results.values.all { it }) {
            loadSims()
            loadPreferences()
        } else {
            val denied = results.filterValues { !it }.keys
            val anyPermanentlyDenied = denied.any {
                !shouldShowRequestPermissionRationale(it)
            }
            if (anyPermanentlyDenied) {
                showOpenSettingsDialog()
            } else {
                Toast.makeText(this, getString(R.string.permission_rationale), Toast.LENGTH_LONG).show()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.btnSave.setOnClickListener { saveSettings() }

        binding.switchEnable.setOnCheckedChangeListener { _, isChecked ->
            PreferencesManager.setEnabled(this, isChecked)
            if (isChecked) startMonitoringService() else stopMonitoringService()
        }

        checkAndRequestPermissions()
    }

    override fun onResume() {
        super.onResume()
        if (allPermissionsGranted()) {
            loadSims()
            loadPreferences()
        }
    }

    private fun checkAndRequestPermissions() {
        val toRequest = buildRequiredPermissions().filter {
            ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
        }

        if (toRequest.isEmpty()) {
            loadSims()
            loadPreferences()
            return
        }

        val needsRationale = toRequest.any { shouldShowRequestPermissionRationale(it) }
        if (needsRationale) {
            AlertDialog.Builder(this)
                .setTitle(R.string.app_name)
                .setMessage(R.string.permission_rationale)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    permissionLauncher.launch(toRequest.toTypedArray())
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
        } else {
            permissionLauncher.launch(toRequest.toTypedArray())
        }
    }

    private fun buildRequiredPermissions(): List<String> = buildList {
        add(Manifest.permission.READ_PHONE_STATE)
        add(Manifest.permission.READ_CALL_LOG)
        add(Manifest.permission.SEND_SMS)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            add(Manifest.permission.READ_PHONE_NUMBERS)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            add(Manifest.permission.POST_NOTIFICATIONS)
        }
    }

    private fun allPermissionsGranted() = buildRequiredPermissions().all {
        ContextCompat.checkSelfPermission(this, it) == PackageManager.PERMISSION_GRANTED
    }

    @SuppressLint("MissingPermission")
    private fun loadSims() {
        val sm = getSystemService(SubscriptionManager::class.java)
        simList = sm.activeSubscriptionInfoList ?: emptyList()

        if (simList.isEmpty()) {
            binding.tvStatus.text = getString(R.string.status_no_sims)
            binding.btnSave.isEnabled = false
            return
        }

        val displayNames = simList.mapIndexed { index, info ->
            val number = info.number?.takeIf { it.isNotBlank() } ?: getString(R.string.sim_unknown_number)
            "SIM ${index + 1}: ${info.displayName} ($number)"
        }

        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, displayNames)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spinnerSim.adapter = adapter

        val savedSubId = PreferencesManager.getSubscriptionId(this)
        val savedIndex = simList.indexOfFirst { it.subscriptionId == savedSubId }
        if (savedIndex >= 0) binding.spinnerSim.setSelection(savedIndex)

        binding.btnSave.isEnabled = true
    }

    private fun loadPreferences() {
        binding.switchEnable.isChecked = PreferencesManager.isEnabled(this)
        binding.editReplyMessage.setText(PreferencesManager.getReplyMessage(this))
    }

    private fun saveSettings() {
        val selectedIndex = binding.spinnerSim.selectedItemPosition
        if (selectedIndex < 0 || selectedIndex >= simList.size) return

        val subId = simList[selectedIndex].subscriptionId
        val message = binding.editReplyMessage.text.toString().trim()

        if (message.isBlank()) {
            Toast.makeText(this, R.string.error_empty_message, Toast.LENGTH_SHORT).show()
            return
        }

        PreferencesManager.setSubscriptionId(this, subId)
        PreferencesManager.setReplyMessage(this, message)
        PreferencesManager.setEnabled(this, binding.switchEnable.isChecked)

        if (binding.switchEnable.isChecked) {
            startMonitoringService()
        } else {
            stopMonitoringService()
        }

        binding.tvStatus.text = getString(R.string.status_saved)
    }

    private fun startMonitoringService() {
        startForegroundService(this, Intent(this, CallStateService::class.java))
        binding.tvStatus.text = getString(R.string.status_service_started)
    }

    private fun stopMonitoringService() {
        stopService(Intent(this, CallStateService::class.java))
        binding.tvStatus.text = getString(R.string.status_service_stopped)
    }

    private fun showOpenSettingsDialog() {
        AlertDialog.Builder(this)
            .setTitle(R.string.app_name)
            .setMessage(R.string.permission_permanently_denied)
            .setPositiveButton(R.string.open_settings) { _, _ ->
                startActivity(
                    Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                        data = Uri.fromParts("package", packageName, null)
                    }
                )
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }
}
